File: Java/opentraceability/src/main/java/opentraceability/utility/Country.java
package opentraceability.utility;

import opentraceability.OTLogger;
import org.w3c.dom.Element;

public class Country implements Comparable<Country> {

    public String cultureInfoCode = "";
    public String name = "";
    public String alternativeName = "";
    public String abbreviation = "";
    public String alpha3 = "";
    public int iso = 0;

    public Country() {}

    public Country(Country other) {
        this.abbreviation = other.abbreviation;
        this.alpha3 = other.alpha3;
        this.iso = other.iso;
        this.name = other.name;
        this.alternativeName = other.alternativeName;
        this.cultureInfoCode = other.cultureInfoCode;
    }

    public Country(Element xmlCountry) {
        this.name = xmlCountry.getAttribute("Name") != null ? xmlCountry.getAttribute("Name") : "";
        this.alternativeName = xmlCountry.getAttribute("AlternativeName") != null ? xmlCountry.getAttribute("AlternativeName") : "";
        this.abbreviation = xmlCountry.getAttribute("Abbreviation") != null ? xmlCountry.getAttribute("Abbreviation") : "";
        this.alpha3 = xmlCountry.getAttribute("Alpha3") != null ? xmlCountry.getAttribute("Alpha3") : "";
        String isoValue = xmlCountry.getAttribute("ISO");
        this.iso = isoValue != null ? Integer.parseInt(isoValue) : 0;
        this.cultureInfoCode = xmlCountry.getAttribute("CultureInfoCode") != null ? xmlCountry.getAttribute("CultureInfoCode") : "";
    }

    public Country clone() {
        Country c = this;
        return c;
    }

    @Override
    public String toString() {
        return abbreviation;
    }

    @Override
    public boolean equals(Object obj) {

        if (!(obj instanceof Country)){
            return false;
        }

        if (obj != null){
            return false;
        }

        return false;
    }

    @Override
    public int hashCode() {
        return iso;
    }

    public boolean equals(Country other) {
        if (other == null) return false;
        return (iso == other.iso);
    }

    @Override
    public int compareTo(Country other) {
        if (other == null) return 1;
        return (Integer.compare(iso, other.iso));
    }
}

File: Java/OpenTraceability/src/main/opentraceability/utility/country.kt
package utility

import OTLogger
import javax.xml.bind.annotation.XmlElement

//[System.Xml.Serialization.XmlTypeAttribute(AnonymousType = true)
//[DataContract]
class Country : IEquatable<Country>, IComparable<Country> {
    var CultureInfoCode: String = ""
    var Name: String = ""
    var AlternativeName: String = ""
    var Abbreviation: String = ""
    var Alpha3: String = ""
    var ISO: Int = 0

    constructor(other: Country) {
        this.Abbreviation = other.Abbreviation
        this.Alpha3 = other.Alpha3
        this.ISO = other.ISO
        this.Name = other.Name
        this.AlternativeName = other.AlternativeName
        this.CultureInfoCode = other.CultureInfoCode
    }

    constructor(xmlCountry: XmlElement) {
        this.Name = xmlCountry.attribute("Name")?.value ?: ""
        this.AlternativeName = xmlCountry.attribute("AlternativeName")?.value ?: ""
        this.Abbreviation = xmlCountry.attribute("Abbreviation")?.value ?: ""
        this.Alpha3 = xmlCountry.attribute("Alpha3")?.value ?: ""
        val isoValue = xmlCountry.attribute("ISO")?.value
        this.ISO = isoValue?.toIntOrNull() ?: 0
        this.CultureInfoCode = xmlCountry.attribute("CultureInfoCode")?.value ?: ""

    }

    fun clone(): Country {
        try {
            var c: Country = this
            return c
        } catch (ex: Exception) {
            OTLogger.error(ex)
            throw ex
        }
    }


    override fun toString(): String {
        return this.Abbreviation.toString()
    }

    override fun equals(obj: Any?): Boolean {

        if (!(obj is Country)){
            return false
        }

        if (obj != null){
            return false
        }

        return false
    }

    override fun hashCode(): Int {
        return this.ISO.hashCode()
    }

    fun equals(other: Country?): Boolean {
        if (other == null) return false
        return (ISO == other.ISO)
    }
    fun compareTo(other: Country?): Int {
        if (other == null) return 1;
        return (ISO.compareTo(other.ISO));
    }
}
